### PR TITLE
PLANET-7521: Migrate Media Block to Audio/Video blocks

### PR DESCRIPTION
**Pull request by mardelnet**
Ref: https://jira.greenpeace.org/browse/PLANET-7521

<hr>

**DESCRIPTION:**
This PR migrates the Media blocks to Audio, Embed or Video blocks.
It also removes the Media blocks from our codebase to prevent users from using them after the migration.

<hr>

**RELATED PRs:**
- https://github.com/greenpeace/planet4-master-theme/pull/2367
- https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/1236

<hr>

**NOTES:**
- As Video blocks connected to YouTube, Vimeo, or Soundcloud elements are automatically turned into a YouTube, Vimeo, or Soundcloud block (which are Embed blocks), this PR migrates those kinds of elements to Embed blocks directly.
- The title of the Media blocks is turned into heading blocks.
- The description of the Media blocks is turned into paragraph blocks.
- The Video poster image of the Media blocks is turned into the poster of the video migrated block.

<hr>

**TESTING INSTRUCTIONS:**
1. Run this PR in your local environment.
2. Identify a post using Media blocks or create a new one with them (at the end of this description, you'll find Media blocks in code format that can be copied and pasted into the post editor).
3. Run the migrator command: `npx wp-env run cli wp p4-run-activator`
4. Check whether the migration worked as expected.
5. To reset the environment database back to the latest default content and repeat the tests, run: `npm run db:reset`. After resetting the database, you'll find some posts using Media blocks, such as the 952, 914, 907, 904, 866 and 810.

<hr>

```
<!-- wp:planet4-blocks/media-video {"video_title":"The Boat and the Bomb","description":"On the night of July 10,1985, the blasts of two limpet mines placed by French secret agents sank a ship protesting against nuclear testing in the Pacific and took the life of onboard photographer Fernando Pereira.\u003cbr\u003e\u003cbr\u003eThis film captures the events leading up to the explosions, as well as the aftermath of what was deemed internationally to be a state-sponsored criminal act of sabotage. As details of the plot – implicating the highest level of the French government – were revealed, waves of outrage rippled across the globe. It fueled the debate about nuclear weapons testing which eventually culminated in the Comprehensive Test Ban Treaty in 1996.","embed_html":"\u003ciframe title=\u0022The Boat and the Bomb (2005) - full length version\u0022 width=\u0022600\u0022 height=\u0022450\u0022 src=\u0022https://www.youtube.com/embed/2uw8tg9_BU4?feature=oembed\u0022 frameborder=\u00220\u0022 allow=\u0022accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\u0022 referrerpolicy=\u0022strict-origin-when-cross-origin\u0022 allowfullscreen\u003e\u003c/iframe\u003e","media_url":"https://youtu.be/2uw8tg9_BU4"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><header><h2 class="page-section-header">The Boat and the Bomb</h2></header><p class="page-section-description">On the night of July 10,1985, the blasts of two limpet mines placed by French secret agents sank a ship protesting against nuclear testing in the Pacific and took the life of onboard photographer Fernando Pereira.<br><br>This film captures the events leading up to the explosions, as well as the aftermath of what was deemed internationally to be a state-sponsored criminal act of sabotage. As details of the plot – implicating the highest level of the French government – were revealed, waves of outrage rippled across the globe. It fueled the debate about nuclear weapons testing which eventually culminated in the Comprehensive Test Ban Treaty in 1996.</p><div><div class="embed-container"><iframe title="The Boat and the Bomb (2005) - full length version" src="https://www.youtube-nocookie.com/embed/2uw8tg9_BU4?feature=oembed" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen=""></iframe></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"Media Block Youtube ID only","description":"","embed_html":"\u003ciframe title=\u0022Work \u0026amp; Study with Batman 🦇 24/7 Deep Cinematic Music for Instant Flow State \u0026amp; Productivity\u0022 width=\u0022600\u0022 height=\u0022338\u0022 src=\u0022https://www.youtube.com/embed/mPhl23eo69g?feature=oembed\u0022 frameborder=\u00220\u0022 allow=\u0022accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\u0022 referrerpolicy=\u0022strict-origin-when-cross-origin\u0022 allowfullscreen\u003e\u003c/iframe\u003e","media_url":"mPhl23eo69g"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><header><h2 class="page-section-header">Media Block Youtube ID only</h2></header><div><div class="embed-container"><iframe title="Work &amp; Study with Batman 🦇 24/7 Deep Cinematic Music for Instant Flow State &amp; Productivity" src="https://www.youtube-nocookie.com/embed/mPhl23eo69g?feature=oembed" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen=""></iframe></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"Media Block Youtube ID + video poster","description":"This is the block description!","video_poster_img":250,"embed_html":"\u003ciframe title=\u0022Japan Cafe Vibe ☕ Lofi Coffee 24/7 beats🍃 Lofi Hip Hop 24/7 / Lofi Music to sleep,relax,study...\u0022 width=\u0022600\u0022 height=\u0022450\u0022 src=\u0022https://www.youtube.com/embed/bRnTGwCbr3E?feature=oembed\u0022 frameborder=\u00220\u0022 allow=\u0022accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\u0022 referrerpolicy=\u0022strict-origin-when-cross-origin\u0022 allowfullscreen\u003e\u003c/iframe\u003e","media_url":"bRnTGwCbr3E","poster_url":"http://www.planet4.test/wp-content/uploads/2018/05/aa1dd182-deep-sea-oil-protest-1024x683.jpg"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><header><h2 class="page-section-header">Media Block Youtube ID + video poster</h2></header><p class="page-section-description">This is the block description!</p><div><div class="embed-container"><iframe title="Japan Cafe Vibe ☕ Lofi Coffee 24/7 beats🍃 Lofi Hip Hop 24/7 / Lofi Music to sleep,relax,study..." src="https://www.youtube-nocookie.com/embed/bRnTGwCbr3E?feature=oembed" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen=""></iframe></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"","description":"","embed_html":"\u003ciframe title=\u0022JORGE TE LO NEGOCIO, SE VINO CON MOWZET, MASLATÓN y DRY SHELLS CORNUDA | HAY ALGO AHÍ | BLENDER\u0022 width=\u0022600\u0022 height=\u0022338\u0022 src=\u0022https://www.youtube.com/embed/rnU6zWE0-4E?feature=oembed\u0022 frameborder=\u00220\u0022 allow=\u0022accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share\u0022 referrerpolicy=\u0022strict-origin-when-cross-origin\u0022 allowfullscreen\u003e\u003c/iframe\u003e","media_url":"https://www.youtube.com/watch?v=rnU6zWE0-4E"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><div><div class="embed-container"><iframe title="JORGE TE LO NEGOCIO, SE VINO CON MOWZET, MASLATÓN y DRY SHELLS CORNUDA | HAY ALGO AHÍ | BLENDER" src="https://www.youtube-nocookie.com/embed/rnU6zWE0-4E?feature=oembed" frameborder="0" allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture; web-share" referrerpolicy="strict-origin-when-cross-origin" allowfullscreen=""></iframe></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"Media Block Vimeo - No description","description":"","video_poster_img":250,"embed_html":"\u003ciframe title=\u0022See Saw\u0022 src=\u0022https://player.vimeo.com/video/859177196?dnt=1\u0026amp;app_id=122963\u0022 width=\u0022600\u0022 height=\u0022338\u0022 frameborder=\u00220\u0022 allow=\u0022autoplay; fullscreen; picture-in-picture; clipboard-write\u0022\u003e\u003c/iframe\u003e","media_url":"https://vimeo.com/channels/bestofstaffpicks/859177196","poster_url":"http://www.planet4.test/wp-content/uploads/2018/05/aa1dd182-deep-sea-oil-protest-1024x683.jpg"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><header><h2 class="page-section-header">Media Block Vimeo - No description</h2></header><div><div class="embed-container"><iframe title="See Saw" src="https://player.vimeo.com/video/859177196?dnt=1&amp;app_id=122963" frameborder="0" allow="autoplay; fullscreen; picture-in-picture; clipboard-write"></iframe></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"","description":"This is the Soundcloud description - No title","video_poster_img":250,"embed_html":"\u003ciframe title=\u0022Torn by ℒund\u0022 width=\u0022600\u0022 height=\u0022400\u0022 scrolling=\u0022no\u0022 frameborder=\u0022no\u0022 src=\u0022https://w.soundcloud.com/player/?visual=true\u0026url=https%3A%2F%2Fapi.soundcloud.com%2Ftracks%2F1882701765\u0026show_artwork=true\u0026maxheight=750\u0026maxwidth=600\u0022\u003e\u003c/iframe\u003e","media_url":"https://soundcloud.com/lund/torn","poster_url":"http://www.planet4.test/wp-content/uploads/2018/05/aa1dd182-deep-sea-oil-protest-1024x683.jpg"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><p class="page-section-description">This is the Soundcloud description - No title</p><div><div class="embed-container"><iframe title="Torn by ℒund" scrolling="no" frameborder="no" src="https://w.soundcloud.com/player/?visual=true&amp;url=https%3A%2F%2Fapi.soundcloud.com%2Ftracks%2F1882701765&amp;show_artwork=true&amp;maxheight=750&amp;maxwidth=600"></iframe></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"","description":"","video_poster_img":276,"embed_html":"[video  src=\u0022http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4\u0022 /]","media_url":"http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4","poster_url":"http://www.planet4.test/wp-content/uploads/2018/05/aa49ade7-gp0stqwmj_pressmedia-communit-policy-1024x684.jpg"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><div class="video-embed"><div class="video-embed-container"><video src="http://commondatastorage.googleapis.com/gtv-videos-bucket/sample/BigBuckBunny.mp4" style="width:100%;height:100%" class="mejs-video-block" poster="http://www.planet4.test/wp-content/uploads/2018/05/aa49ade7-gp0stqwmj_pressmedia-communit-policy-1024x684.jpg" controls preload="metadata"></video></div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"","description":"This is the MP3 block description - No title","video_poster_img":276,"embed_html":"[audio src=\u0022https://samplelib.com/lib/preview/mp3/sample-15s.mp3\u0022 /]","media_url":"https://samplelib.com/lib/preview/mp3/sample-15s.mp3","poster_url":"http://www.planet4.test/wp-content/uploads/2018/05/aa49ade7-gp0stqwmj_pressmedia-communit-policy-1024x684.jpg"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><p class="page-section-description">This is the MP3 block description - No title</p><div><div class="embed-container">[audio src="https://samplelib.com/lib/preview/mp3/sample-15s.mp3" /]</div></div></section>
<!-- /wp:planet4-blocks/media-video -->

<!-- wp:planet4-blocks/media-video {"video_title":"Media Block wav file - No description","description":"","video_poster_img":276,"embed_html":"[audio src=\u0022https://www2.cs.uic.edu/~i101/SoundFiles/PinkPanther60.wav\u0022 /]","media_url":"https://www2.cs.uic.edu/~i101/SoundFiles/PinkPanther60.wav","poster_url":"http://www.planet4.test/wp-content/uploads/2018/05/aa49ade7-gp0stqwmj_pressmedia-communit-policy-1024x684.jpg"} -->
<section class="block media-block wp-block-planet4-blocks-media-video"><header><h2 class="page-section-header">Media Block wav file - No description</h2></header><div><div class="embed-container">[audio src="https://www2.cs.uic.edu/~i101/SoundFiles/PinkPanther60.wav" /]</div></div></section>
<!-- /wp:planet4-blocks/media-video -->
```

### CODE REVIEW

**Comment by planet-4** [comment]
### Test instance is ready :rocket:

:new_moon: [tavros](https://www-dev.greenpeace.org/test-tavros) | [admin](https://www-dev.greenpeace.org/test-tavros/wp-admin/) | [blocks report](https://www-dev.greenpeace.org/test-tavros/wp-admin/?page=plugin_blocks_report) | [CircleCI](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-tavros) | [composer-local.json](https://github.com/greenpeace/planet4-test-tavros/blob/main/composer-local.json)

:watch: 2024.08.23 11:45:50

**Review by Osong-Michael — COMMENTED** [review]
The PR looks good so far, just one comment at the moment. 
Converting the _Media_ block with a SoundCloud URL to the equivalent _Audio_ block, shows fine in the frontend but in the editor there is an issue.

soundcloud

**Comment by mardelnet** [comment]
> The PR looks good so far, just one comment at the moment. Converting the _Media_ block with a SoundCloud URL to the equivalent _Audio_ block, shows fine in the frontend but in the editor there is an issue.

Fixed, @Osong-Michael 
Thanks for the feedback!

**Review by Osong-Michael — APPROVED** [review]
👍

File: src/Migrator.php [hunk 1]
```diff
@@ -32,6 +32,7 @@
 use P4\MasterTheme\Migrations\M028MovePageHeaderSideBarOptions;
 use P4\MasterTheme\Migrations\M029RemoveTemplateEditorOption;
 use P4\MasterTheme\Migrations\M030RemovePurgeOnFeatureChangeOption;
+use P4\MasterTheme\Migrations\M033MigrateMediaBlockToAudioVideoBlock;
```
Comment (comzeradd) [inline src/Migrator.php:35]: Shouldn't that prefixed as `M031` ?

**Comment by comzeradd** [comment]
I think we should also remove the Media block at the same time, to avoid people adding new instance of that after the migration runs

**Comment by mardelnet** [comment]
> I think we should also remove the Media block at the same time, to avoid people adding new instance of that after the migration runs

I have already created PRs to tackle that, @comzeradd:
- https://github.com/greenpeace/planet4-master-theme/pull/2367
- https://github.com/greenpeace/planet4-plugin-gutenberg-blocks/pull/1236

File: src/Migrator.php (hunk 1 above)
Comment (mardelnet) [reply src/Migrator.php]: Done!

**Comment by mardelnet** [comment]
Hey @Osong-Michael, would you mind re-checking this PR?
The following changes were introduced:
1. The Media block is removed to prevent editors from using it.
2. The title and description of the Media blocks are no longer converted into video/audio captions but moved to head and paragraph blocks. So, after the migration, you'll see the Media block turned into a Group block containing a head, a paragraph and a video/audio/embed block.

**Comment by sagarsdeshmukh** [comment]
Hi @mardelnet , To track the migration script execution, we should include the migration script name and post ID along with each post migration status. 
This information will help in tracking and analyzing the post-migration results. I have added these details in the [campaign covers](https://github.com/greenpeace/planet4-master-theme/pull/2349/files#diff-8a2cfe9830302d03307b1ac039eb67c773abc15030adcbb9e3eece0708a738c6R81) and [split 2 columns](https://github.com/greenpeace/planet4-master-theme/pull/2350/files#diff-caf19d90fc201d2db867d04cd8f157f6bb37183e6113012af080bec25d252c5cR44) PR, making it easy to verify which posts have been migrated in the [post-deployment](https://app.circleci.com/pipelines/github/greenpeace/planet4-test-titan/801/workflows/24380812-ed7a-4703-a282-406b071778cb/jobs/2939?invite=true#step-108-7176_16) script.

Your feedback on the above would be most welcome.

File: src/Migrations/M031MigrateMediaBlockToAudioVideoBlock.php [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+        $block['innerHTML'] =
+        '<div class="wp-block-group">
+
+
+
+
+
+        </div>';
+
+        // IMPORTANT: DO NOT MODIFY THIS FORMAT!
```
Comment (sagarsdeshmukh) [inline src/Migrations/M031MigrateMediaBlockToAudioVideoBlock.php:175]: I tried a different approach to transform a block in a [split 2 columns block](https://github.com/greenpeace/planet4-master-theme/pull/2350/files#diff-caf19d90fc201d2db867d04cd8f157f6bb37183e6113012af080bec25d252c5cR95-R139) migration, we may use similar solution here as well.

Comment (mardelnet) [reply src/Migrations/M031MigrateMediaBlockToAudioVideoBlock.php:175]: Thanks @sagarsdeshmukh 
I have a slightly different opinion about the approach.
Please, check my comment [here](https://github.com/greenpeace/planet4-master-theme/pull/2350#discussion_r1751540000), and let's discuss it further.

**Comment by comzeradd** [comment]
The migration works great. I tested it with a few use cases and all work great after the migration is executed. But two questions:

1. Shouldn't we also add audio and video blocks in [BlockSettings](https://github.com/greenpeace/planet4-master-theme/blob/v1.296.0/src/BlockSettings.php) as part of this PR? So that users can use the blocks after the migration.
2. I noticed that the native video block is embedding an iframe with fixed width (usually 500px) while our current custom implementation is full width. Maybe something to check at a follow up ticket, although I suspect that this is actually better in terms of responsiveness and [CLS](https://web.dev/articles/cls).

**Comment by mardelnet** [comment]
Thanks @comzeradd

>     1. Shouldn't we also add audio and video blocks in [BlockSettings](https://github.com/greenpeace/planet4-master-theme/blob/v1.296.0/src/BlockSettings.php) as part of this PR? So that users can use the blocks after the migration.
Done! Check it [here](https://github.com/greenpeace/planet4-master-theme/pull/2357/commits/d8ff4f20d30996473268318e51eca308de23dc2a).
 
>     2. I noticed that the native video block is embedding an iframe with fixed width (usually 500px) while our current custom implementation is full width. Maybe something to check at a follow up ticket, although I suspect that this is actually better in terms of responsiveness and [CLS](https://web.dev/articles/cls).
I agree this can be worked on a separate ticket.

**Review by comzeradd — APPROVED** [review]
Nice work 🍻 

_(don't forget to squash your commits)_

**Pull request merged**